torch_xla/distributed/xrt_init.py

```diff
@@ -0,0 +1,241 @@
+        # the gpu used by this process
+        os.environ['CUDA_VISIBLE_DEVICES'] = str(local_rank)
+
+
+def _init_xrt_context(master_addr=None, master_port=None, store=None):
+    '''
+    Initializes the XLA device depending on the kind of the device. Or is a no-op if init_xrt_context
+    has already been called.
+
+    Keyword Arguments:
+    master_addr -- int(SomeAddrNumber) (default tries to get addr from
+                    environment variable, used to setup TCPStore)
+                    Not needed if store is provided.
+    master_port -- int(SomePortNumber) (default tries to get port from
+                    environment variable, used to setup TCPStore)
+                    Not needed if store is provided.
+    store       -- A TCPstore object (default None) -- If None a TCPStore object
+                    will be setup for you
+    '''
+    global _INIT_XRT_ALREADY_CALLED
+
+    if _INIT_XRT_ALREADY_CALLED:
+        return
+
+    # Call this in the actual test case, to work with torch/xla workers
+    rank = int(os.environ['RANK'])
+    local_rank = int(os.environ['LOCAL_RANK'])
+    world_size = int(os.environ['WORLD_SIZE'])
+    group_rank = int(os.environ['GROUP_RANK'])
+    local_world_size = int(os.environ['LOCAL_WORLD_SIZE'])
+
+    if master_addr is None:
+        master_addr = os.environ['MASTER_ADDR']
+
+    if master_port is None:
+        master_port = os.environ['MASTER_PORT']
+
+    try:
+        _, dev_kind = _get_devices_per_worker()
+    except RuntimeError as e:
+        # Defaulting to TPU if no env variable is set
+        dev_kind = 'TPU'
+    os.environ.pop(xenv.TPU_CONFIG, None)
+    os.environ.pop(xenv.TPU_NUM_DEVICES, None)
+    os.environ.pop(xenv.GPU_NUM_DEVICES, None)
+
+    # This is required if we want to dynamically grab free ports.
+    # Useful in shared settings when we cannot predetermine what ports are taken.
+    is_server = True if rank is '0' else False
+    global _TCP_STORE
+    if store is None:
+        assert master_addr is not None
+        assert master_port is not None
+        _TCP_STORE = dist.TCPStore(master_addr, int(master_port), world_size,
+                                   is_server)
+    else:
+        _TCP_STORE = store
+
+    node_list = None
+
+    if dev_kind == 'TPU':
+        tpu_config_port = _set_tpu_xrt_envs(local_rank, rank, group_rank,
+                                            local_world_size, world_size)
+    elif dev_kind == 'GPU':
+        set_xrt_envs(dev_kind, world_size, rank, local_rank)
+        _create_devices(dev_kind, world_size)
+        _setup_workers(world_size, rank, False, node_list)
+
+    _set_mesh_config(rank)
+
+    if dev_kind == 'TPU':
+        _setup_nccl_service(rank)
+        _set_neuron_envs(rank, world_size, local_world_size)
+
+        total_nodes = world_size // local_world_size
+        if local_rank == 0:
+            local_env = os.environ.copy()
+            subprocess.Popen([
```

Comment:
Do you need to use subprocess here? Can you call [`xrt_run_server.run_service`](https://github.com/pytorch/xla/blob/590dee5d08ff4cf3ba72eab306b470e22bd08e16/torch_xla/core/xrt_run_server.py#L25) directly with [`multiprocessing`](https://docs.python.org/3/library/multiprocessing.html) instead?

Reply:
Will, we basically follow the same template as done in xrt_run_server.py using subprocess.Popen. Essentially, its the same server invocation. However, what we do extra is spawn up an extra "watch" process that polls if the original process is around or got killed(This happens when Slurm terminates the workers and Slurm is not aware of this server, So, the watch process does this job).